src/models/state/wallet/wallet_state.rs

```diff
@@ -3662,4 +3865,331 @@ mod tests {
+            let wallet_secret = WalletSecret::new_random();
+            let data_dir = mock_data_dir();
+            let rusty_wallet_database = mock_rusty_wallet_database().await;
+
+            // activate scan mode by setting --scan-blocks, so num future keys
+            // will assume its default value
+            let cli_args = Args {
+                scan_blocks: Some(0u64..=10),
+                ..Default::default()
+            };
+            let wallet_state = WalletState::new_from_wallet_secret_and_database(
+                rusty_wallet_database,
+                &data_dir,
+                wallet_secret.clone(),
+                &cli_args,
+                false,
+                false,
+            )
+            .await;
+            let scan_mode = wallet_state.scan_mode.unwrap();
+            assert!(scan_mode.num_future_keys() > 0);
+            assert!(scan_mode.num_future_keys() < 1_000_000);
+        }
+
+        #[tokio::test]
+        async fn block_height_range_check_agrees_with_interval_membership() {
+            let wallet_secret = WalletSecret::new_random();
+            let data_dir = mock_data_dir();
+            let rusty_wallet_database = mock_rusty_wallet_database().await;
+
+            let lb = 10;
+            let ub = 20;
+
+            // activate scan mode by setting --scan-blocks, so num future keys
+            // will assume its default value
+            let cli_args = Args {
+                scan_blocks: Some(lb..=ub),
+                ..Default::default()
+            };
+            let wallet_state = WalletState::new_from_wallet_secret_and_database(
+                rusty_wallet_database,
+                &data_dir,
+                wallet_secret.clone(),
+                &cli_args,
+                false,
+                false,
+            )
+            .await;
+            let scan_mode = wallet_state.scan_mode.unwrap();
+
+            for h in (lb - 5)..(ub + 5) {
+                assert_eq!(
+                    (lb..=ub).contains(&h),
+                    scan_mode.block_height_is_in_range(BlockHeight::from(h)),
+                );
+            }
+        }
+
+        /// Test scan mode.
+        ///
+        /// In rough terms, this test verifies that importing a wallet followed
+        /// by booting the node in scan mode will recover UTXOs sent to it.
+        ///
+        /// Specifically:
+        ///  - Alice is recipient of a transaction in block 1, but the address
+        ///    for this transaction uses derivation index 20, which is in the
+        ///    future.
+        ///  - If Alice does nothing special, she does not catch the UTXO.
+        ///  - If Alice activates scan mode with the right parameters, she does
+        ///    catch the UTXO.
+        ///  - In the last case, her derivation counter is updated accordingly.
+        #[traced_test]
+        #[tokio::test]
+        async fn test_recovery_on_imported_wallet() {
```

Comment:
Pretty good test that I think captures a very big part of the logic. This seems like a very fine-grained test to me. Cool!